### ISSUE TITLE
[Discussion] By arrow key movement, the cursor should be able to reach the inline end.

### ISSUE THREAD

**Issue by zhujinxuan** [ianstormtaylor/slate#1610]
#### Do you want to request a *feature* or report a *bug*?

feature
Relate to https://github.com/ianstormtaylor/slate/pull/1582

#### What's the current behavior?
If I am not wrong, the `onSelect` skip the inline end because the mouse click in some browsers cannot catch the inline end.  But the logic of key input, which is decided by `packages/slate` slate core codebase, is consistent crossing browsers.  We should normalize the `selection` when and only when the mouse click happens.

I think we should provide `is{Something}VisiblyTheSame` for `on{Event}`.  These methods are designed to 
  1. If the computed slate variable is not significantly different from the existing range in the slate value, then do not normalize
  2. if normalize, we would prefer the range closer to the existing range in slate value
  3. otherwise, use the normalized range

Then we still have consistent behavior between browsers, and we can have the cursor at the end of inline to insert and delete.

**Comment by ianstormtaylor** [comment]
As far as I remember, it's not because browsers weren't able to handle clicks, it was that they changed any selections at the end of inlines to being in the following text nodes instead. So I'm not sure we can do this.

If someone wants to do the cross-browser research to reach a solution that doesn't require us to normalize the selection though I'm all for it.

**Comment by zhujinxuan** [comment]
Do you mean, in some browsers, the value of native DOM selection
 ```
   let range = findRange(native, value)
```
Always get the `range` after inline?

Sorry about my previous poor explanation.  

My plan is, if `let range = findRange(native, value)` is visibly different from the `value.selection`, then we can inspect that the use do things like mouse click or any other operations that not directly handled by slate-core, then we should update the `value.selection`

If we find that `value.selection` is visibly indifferent from `range`, we can expect that the
1. the value.selection is processed by `input`, `arrowKey` or anythings alike well handled by slate-core
2. the user clicks at the cursor for some reason, but this click should not have significant result.

By visibly indifference I mean thing like
```
<link> something<curosr/></link>another
```
is visibly indifferent to 
```
<link> something</link><curosr/>another
```
and is also visibly indifferent to 
```
<link> something<anchor/></link><focus/>another
```

Then that is my idea of handling cross-browsers compatibility on this question.  If it is reasonable to keep the slate's `value.selection`, keep the `value.selection` because slate-core is cross-browsers compatible.

**Comment by ianstormtaylor** [comment]
Last I checked, Chrome/Safari explicitly prevented selections from being:

```
<link>something<cursor/></link>another
```

They would automatically change them to:

```
<link>something</link><cursor/>another
```

If I remember correctly.

**Comment by zhujinxuan** [comment]
Would you mind the the small inconsistency between browser and `editor.value.selection` like that?

When the `editor.value` is 
```
<link>something<cursor/></link>another
```
but the native DOM is 
```
<link>something</link><cursor/>another
```
Then the `editor.value` is still 
```
<link>something<cursor/></link>another
```

I think users cannot feel the inconsistency from the two, and by that, we can move more control to the slate-core, and avoid the 'select inline and insertText bug'

**Comment by ianstormtaylor** [comment]
@zhujinxuan I think the issue is that this would leave open inconsistencies between browsers then, since your cursor might be in two different spots depending on which browser you were in?

I'm open to the idea, but it's going to require someone doing research. I'm not going to accept/debate changes without proof that they are cross-browser compatible. Or that the issue isn't an issue for other reasons.

**Comment by zhujinxuan** [comment]
@ianstormtaylor Great. I will test operations on Mac's firefox, chrome and safari.

**Comment by zhujinxuan** [comment]
Unable to move to end by arrow however, but it seems it is consistency between FF, Chrome and Safari under Mac.

Arrow and Select All operations:
Chrome
![](http://g.recordit.co/OGLpe4B20d.gif)
Firefox 
![](http://g.recordit.co/uhdpeBUcXG.gif)
Safari
![](http://g.recordit.co/uloe3mgQKO.gif)

By mouse select:

**Comment by urugator** [comment]
Hi, not sure if it helps somehow, but I tried to briefly test the default behavior of `contenteditable` on Chrome, what I found out:
- it doesn't matter whether the inline is followed by some other inline or not, the behavior is always the same
- the selection is always in the same node as the character directly preceeding the selection (this is even indicated by the cursor having the same styling as the character before) - so actually, the selection is always in the "previous", not in the "next" node...
- THE TWIST: even though the selection is always inside "the previous" inline, whenever user inserts (by typing/pasting) any content, the selection jumps right after the current node prior to inserting the content (this seems like bug to me)

In conclusion:
- selection is always in previous inline
- inserting content causes the selection to move outside prior to inserting content

Probably naive idea ... could we preventDefault all insertions and perform them manually?

**Comment by urugator** [comment]
In firefox the behavior depends on how the selection was obtained:
- mouse click: depends on which inline the user actually clicked (the cursor can end up in both)
- mouse selection: same as click - it depends where mouseUp was performed (this is a little bit weird because you may have a selection like aaa[bbb] and end up in "aaa") 
- arrow keys: depends on direction - the selection ends up in the node closer to previous selection:
`|<aaa><bbb> -> <aaa|><bbb>`
`<aaa><bbb>| <- <aaa><|bbb>`

**Comment by urugator** [comment]
I removed that `onSelect` "normalizing" code in `after.js` and there are these two inconsitencies between Chrome and FF:

1. When user move selection backward with arrowleft:  
Ch: `<inline><t|ext>` => (arrowleft) => `<inline|><text>` (at the end of inline)  
FF: `<inline><t|ext>` => (arrowleft) => `<inline><|text>` (in text at offset 0) 
(strangely enought moving backward with backspace is consistent - in text at offset 0)

2. On Chrome it's impossible to place the selection at the start of the text node by clicking, user always end up at the end of the inline: `<inline|><text>`
I think this limitation isn't important to us ... it doesn't mean that the selection can't be placed at the start of the text node, it simply means it's not possible to do that by clicking... 

I can't comment on Safari, but I think that if we could somehow solve the "1." (to match the FF behavior), then there is no issue...

**Comment by urugator** [comment]
So I simply did following in `after.js`:
```javascript
// after.js onKeyDown
if (HOTKEYS.COLLAPSE_CHAR_BACKWARD(event)) {
  event.preventDefault()        
  return change.collapseCharBackward()
  /*
  const { document, isInVoid, previousText, startText } = value
  const isPreviousInVoid =
    previousText && document.hasVoidParent(previousText.key)
  if (isInVoid || isPreviousInVoid || startText.text == '') {
    event.preventDefault()        
    return change.collapseCharBackward()
  }
  */
}
```
Which seems to solve the arrowleft problem.

However there is an issue with pasting... it works consistently (FF and Chrome), but imho in a wrong way ... it doesn't respect the selection (end of inline), but puts the inserted content always to next text node:
`<inline|><text>` (paste "AAA") `<inline><AAAtext>`
This probably happens somewhere in `insertFragmentAtRange` any hints why and where?

**Comment by ianstormtaylor** [comment]
@urugator interesting! Great research. 

For the inserting fragment, it sounds like we need to handle the case on the fragment containing a single text node separately without splitting.

**Comment by miracle2k** [comment]
So, I am struggling with this issue, because I have a case where I want the user to type at the end of an inline, expanding the inline with additional keystrokes. Currently, this seems impossible.

We already saw a lot of examples of browser behavior, but it is been a year, so I will describe what I am seeing now. Using the [Links example](https://www.slatejs.org/#/links), on OSX with the current slate master, Firefox and Chrome behave suchly:

- At the **inline end**, the cursor is always placed **outside** of the inline.
- At the **inline start*, the cursor can be either outside or inside, depending on the direction you came from with the arrow keys.

What seems to happen here is that when advancing to the inline end via "arrow right", Slate.js is initially placing the cursor inside of the inline, at the end. Then, the `<Content>` component's `componentDidUpdate` is called, starting a selection sync, causing `removeAllRanges` to be called, which triggers `onNativeSelectionChange`, which causes [this code](https://github.com/ianstormtaylor/slate/blob/9694b228464d8b4d874074496a4c0a50f6ec4614/packages/slate-react/src/utils/set-selection-from-dom.js#L49-L67) in `setSelectionFromDOM` to normalize the cursor to *after* the inline.

- At the **inline end**, the cursor is always placed **inside** of the inline. Actually, the logic just described tries to move the cursor **outside** of the inline, and in the Slate.js model it is, but when typing, text is inserted **inside**.
- At the **inline start**, the cursor is always placed **outside** of the inline. Here too, the cursor is first **inside**, for example a Slate path of `offset={undefined} path={[0, 1, 0]}`, and is then immediately adjusted to be **outside** (`offset={62} path={[0, 0]}`). I haven't figured out which code path is responsible here.

Why does a bold text not suffer from these issues? I can type at the end of a bold text just fine, and what I type will also be bold. I know for Slate.js bold text is a mark, and thus we never leave the text and cannot even model the concept of being inside/outside the bold text. But in DOM/contenteditable, it is a span, and should therefore suffer from the same inconsistencies that we are trying to normalize here for inlines, no?

The "depending on which side the cursor comes from, it will be inside or outside the inline" approach sounds the most sensible to me, though in this case, I really need the opposite of the current behaviour - I want to force the cursor to always be inside of an inline. I am going to try to run a patched version, but I do think putting more control into the hands of the developer about how the cursor placement is a good thing. 

I'd be willing to do the research, but I am just not sure what the design parameters are here. The behaviour already is inconsistent between Safari and the others. Is we let the developer place the cursor on either side of an inline boundary, regardless of where it is in the DOM, what kind of problems would we expect?

**Comment by klis87** [comment]
I also had issue with it, and I solved it by normalizing given inline nodes to always end with space, sth like:
```js
  schema: {
    inlines: {
      someNode: {
        text: string => !string.slice(string.length - 1, string.length).trim(),
        normalize: (editor, error) => {
          if (error.code === 'node_text_invalid') {
            editor.insertNodeByKey(
              error.node.key,
              error.node.nodes.size,
              Text.create({ text: ' ' }),
            );
          }
        },
      },
    },
  },
```

**Comment by miracle2k** [comment]
I also made a fair amount of progress with the following approach:

```
onCommand(command: any, editor: Editor, next: any) {
      const { type, args } = command
      if (type === 'insertTextAtRange') {
        const [text, range] = args;
      }

      const textNode = editor.value.document.getNode(range.start.path!);
      if (!textNode) {
        return false;
      }

      const isAtBeginningOfText = !range.start.offset;
      const isAtEndOfText = range.start.offset === textNode.text.length;

      // Force cursor to insert at the end of the word, inside of it.
      if (isAtBeginningOfText) {
        let prev = editor.value.document.getPreviousNode(textNode.key);
        if (prev && prev.object === 'inline') {
          editor.insertNodeByKey(
            prev.key,
            prev.nodes.size,
            Text.create({ text: text }),
          );
          return true;
        }
      }

      if (isAtEndOfText) {
        let next = editor.value.document.getNextNode(textNode.key);
        if (next && next.object === 'inline') {
          editor.insertNodeByKey(
            next.key,
            0,
            Text.create({ text: text }),
          );
          editor.moveForward(1);
          return true;
        }
      }
      
      // We did not handle it.
      return next();
    }
  }
```

It checks if the cursor is in one of these critical locations before or after an inline, then inserts the character at the desired place.

**Comment by miracle2k** [comment]
Has anyone had a look at the new Wordpress "Project Gutenberg" editor? It is written in React, and they have a very explicit inline cursor behaviour: You can put the cursor on both sides of the inline boundary, both at the word beginning and end, and you have to press ARROW RIGHT twice to proceed. You can try it here:

https://wordpress.org/gutenberg/

Also for reference, the code seems to be here: https://github.com/WordPress/gutenberg/tree/master/packages/block-editor

**Comment by ianstormtaylor** [comment]
Closing in favor of https://github.com/ianstormtaylor/slate/issues/3148
